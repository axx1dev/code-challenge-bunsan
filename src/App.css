.second-level {
  display: inline-flex;
}

.btn-outline-success, .btn-outline-danger {
  margin-top: 4px;
  margin-bottom: 10px;
  margin-left: 10px;
  border-radius: 20px !important;
}

.third-level {
  display: flex;
}