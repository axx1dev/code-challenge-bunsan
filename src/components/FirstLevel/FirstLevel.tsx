import { PropsLevel } from "../../interfaces/index"

const FirstLevel = ({title, allocation, classFather, classSon}: PropsLevel) => {
  return (
    <h2 className={classFather}>-<span className={classSon}>{`${title} a $${allocation}`}</span></h2>
  )
}

export default FirstLevel