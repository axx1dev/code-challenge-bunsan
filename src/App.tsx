import { useState, useEffect } from "react";
import FirstLevel from "./components/FirstLevel/FirstLevel";
import OverlayTrigger from "react-bootstrap/OverlayTrigger";
import Popover from "react-bootstrap/Popover";
import { Button } from "react-bootstrap";
import { Child } from "./interfaces/index"
import "bootstrap/dist/css/bootstrap.css";
import './App.css';

function App() {
  const [deleteItem, setDeleteItem] = useState<Child>();
  const [type, setType] = useState("");
  const [allocation, setAllocation] = useState(0);
  let total = 0;

  const [dataOfEmployes, setDataEmployes] = useState({
    type: "Manager A",
    allocation: 300,
    childs: [
      {
        type: "Manager B",
        allocation: 300,
        childs: [
          {
            type: "Developer",
            allocation: 1000,
          },
          {
            type: "QA Tester",
            allocation: 500,
          },
        ],
      },
      {
        type: "Manager C",
        allocation: 300,
        childs: [],
      },
    ],
  });

  const addItemtoArray = (position: number) => {
    dataOfEmployes.childs[position].childs.push({
      type: type,
      allocation: allocation,
    });

    setDataEmployes(dataOfEmployes);
    setType("");
    setAllocation(0);
  };

  const deleteIteminArray = (positionChild: number, positionFather: number) => {
    setDeleteItem(dataOfEmployes.childs[positionFather].childs[positionChild]);
    dataOfEmployes.childs[positionFather].childs.splice(positionChild, 1);
    setDataEmployes(dataOfEmployes);
  };

  useEffect(() => {}, [dataOfEmployes, type, allocation, deleteItem]);

  const getTotalAmount = (data: Child) => {
    total += data.allocation;
    data.childs?.map((item: any) => {
      getTotalAmount(item);
    });
    return total;
  };

  const popover = (index: number) => (
    <Popover id="popover-basic">
      <Popover.Header as="h3">Add role position</Popover.Header>
      <Popover.Body>
        
        <div className="input-group mb-3">
          <input
            type="text"
            className="form-control"
            placeholder="Name"
            aria-label="Type"
            aria-describedby="basic-addon1"
            onChange={(e) => setType(e.target.value)}
          />
          <input
            type="number"
            className="form-control"
            placeholder="Allocation"
            aria-label="Allocation"
            aria-describedby="basic-addon1"
            onChange={(e) => setAllocation(parseInt(e.target.value))}
          />
          <Button onClick={() => addItemtoArray(index)}>save</Button>
        </div>
      </Popover.Body>
    </Popover>
  );

  return (
    <div className="App">
      <FirstLevel
        title={dataOfEmployes.type}
        allocation={dataOfEmployes.allocation}
        classSon="badge text-bg-dark"
      />
      {dataOfEmployes?.childs.map((item, index) => (
        <div key={index}>
          <div className="second-level">
            <FirstLevel
              title={item.type}
              allocation={item.allocation}
              classFather="ms-3"
              classSon="badge text-bg-primary"
            />
            <OverlayTrigger trigger="click" placement="right" overlay={popover(index)}>
              <a aria-label="add-btn"><Button variant="outline-success">+</Button></a>
            </OverlayTrigger>
          </div>
          

          {item?.childs?.map((subItem, indexChild) => (
              <div className="third-level container" key={`${index}_${indexChild}`}>
                <FirstLevel
                  title={subItem.type}
                  allocation={subItem.allocation}
                  classFather="ms-5"
                  classSon="badge text-bg-info"
                />
                <Button variant="outline-danger" onClick={() => deleteIteminArray(indexChild, index)}>delete</Button>
              </div>
          ))}
        </div>
      ))}

      <h2>Total: {getTotalAmount(dataOfEmployes)}</h2>
    </div>
  );
}

export default App;
