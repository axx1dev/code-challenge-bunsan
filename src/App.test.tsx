import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import App from './App';
import FirstLevel from "./components/FirstLevel/FirstLevel"
import renderer from 'react-test-renderer';

test('renders Total: word', () => {
  render(<App />);
  const word = screen.getByText(/Total:/i);
  expect(word).toBeInTheDocument();
});

it('rendered FirstLevel Component snapshot', () => {
  const component = renderer.create(
    <FirstLevel title='Manager X' allocation={2300} classFather='ms-5' classSon='badge text-bg-info' />,
  );

  let result = component.toJSON();
  expect(result).toMatchSnapshot();
});

test('Simulate click to add Role Position', async () => {
  render(<App/>);
  const btn = await waitFor(() => screen.getAllByText('+'))
  fireEvent.click(btn[0])
  expect(screen.getByText('Add role position')).toBeInTheDocument();
});