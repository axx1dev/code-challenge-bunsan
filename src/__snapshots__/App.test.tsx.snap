// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendered FirstLevel Component snapshot 1`] = `
<h2
  className="ms-5"
>
  -
  <span
    className="badge text-bg-info"
  >
    Manager X a $2300
  </span>
</h2>
`;
